flow/activities/maintenance_activity.go

```diff
@@ -0,0 +1,292 @@
+package activities
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"go.temporal.io/sdk/activity"
+	"go.temporal.io/sdk/client"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/PeerDB-io/peer-flow/alerting"
+	"github.com/PeerDB-io/peer-flow/generated/protos"
+	"github.com/PeerDB-io/peer-flow/model"
+	"github.com/PeerDB-io/peer-flow/peerdbenv"
+	"github.com/PeerDB-io/peer-flow/shared"
+	"github.com/PeerDB-io/peer-flow/shared/telemetry"
+)
+
+const (
+	mirrorStateBackup   = "backup"
+	mirrorStateRestored = "restore"
+)
+
+type MaintenanceActivity struct {
+	CatalogPool    *pgxpool.Pool
+	Alerter        *alerting.Alerter
+	TemporalClient client.Client
+}
+
+func (a *MaintenanceActivity) GetAllMirrors(ctx context.Context) (*protos.MaintenanceMirrors, error) {
+	rows, err := a.CatalogPool.Query(ctx, `
+	select distinct on(name)
+	  id, name, workflow_id,
+	  created_at, coalesce(query_string, '')='' is_cdc
+	from flows
+	`)
+	if err != nil {
+		return &protos.MaintenanceMirrors{}, err
+	}
+
+	maintenanceMirrorItems, err := pgx.CollectRows(rows, func(row pgx.CollectableRow) (*protos.MaintenanceMirror, error) {
+		var info protos.MaintenanceMirror
+		var createdAt time.Time
+		err := row.Scan(&info.MirrorId, &info.MirrorName, &info.WorkflowId, &createdAt, &info.IsCdc)
+		info.MirrorCreatedAt = timestamppb.New(createdAt)
+		return &info, err
+	})
+	return &protos.MaintenanceMirrors{
+		Mirrors: maintenanceMirrorItems,
+	}, err
+}
+
+func (a *MaintenanceActivity) getMirrorStatus(ctx context.Context, mirror *protos.MaintenanceMirror) (protos.FlowStatus, error) {
+	return shared.GetWorkflowStatus(ctx, a.TemporalClient, mirror.WorkflowId)
+}
+
+func (a *MaintenanceActivity) WaitForRunningSnapshots(ctx context.Context) (*protos.MaintenanceMirrors, error) {
```

Comment:
I'm thinking that it's probably simpler and better to fail entering maintenance mode than wait for an unknown duration of time for running snapshots to finish, we can process upgrades for stuff that enters maintenance mode and try these instances later

Reply:
We have a background alerter for these uses cases where we wait, and it might not be a good idea to fail entering maintenance mode: the whole idea is for upgrades to be automatic.

Reply:
> and try these instances later
This is what maintenance mode will also aim to do